model_compression_toolkit/data_generation/keras/optimization_utils.py

```diff
@@ -0,0 +1,436 @@
+
+        Args:
+            images (tf.Tensor): A tensor containing the input images.
+            optimizer (Any): An optimizer responsible for updating the image parameters during optimization.
+            scheduler (Any): A scheduler responsible for adjusting the learning rate of the optimizer over time.
+            initial_lr (float): The initial learning rate used by the optimizer.
+        """
+        self.images = images
+        self.optimizer = optimizer(lr=initial_lr)
+        self.scheduler = scheduler(optim_lr=self.optimizer)
+
+
+class KerasAllImagesStatsHolder(AllImagesStatsHolder):
+    """
+    An extension of AllImagesStatsHolder specifically for Keras models.
+    Stores activation statistics for all image batches. It offers an organized mechanism for retaining mean,
+    second-moment, and standard deviation statistics corresponding to the activations of each layer.
+
+    Responsible for collecting and storing activation statistics across all batches of images.
+    It stores a list 'batches_stats_holder_list' of 'BatchStatsHolder's. Each `BatchStatsHolder` instance in
+    the `batches_stats_holder_list` is responsible for storing statistics for a specific batch,
+    specified by "batch_index".
+    """
+
+    def get_batches_stats_holder_list(self) -> List[BatchStatsHolder]:
+        """
+        Get a list of BatchStatsHolder objects.
+
+        Returns:
+            List[BatchStatsHolder]: A list of BatchStatsHolder objects.
+        """
+        return [KerasBatchStatsHolder(self.mean_axis) for _ in range(self.n_batches)]
+
+    def update_batch_stats(self,
+                           batch_stats_holder: BatchStatsHolder,
+                           input_imgs: tf.Tensor,
+                           activation_extractor: KerasActivationExtractor):
+        """
+        Update the batch statistics for a given batch.
+
+        Args:
+            batch_stats_holder (BatchStatsHolder): batch stats holder.
+            input_imgs (tf.Tensor): the input images for which to calculate the statistics.
+            activation_extractor (ActivationsExtractor): the activation extractor object.
+        """
+        batch_stats_holder.clear()
+        batch_stats_holder.calc_bn_stats_from_activations(input_imgs=input_imgs,
+                                                          activation_extractor=activation_extractor)
+
+    def get_stats(self,
+                  batch_stats_holder: BatchStatsHolder,
+                  layer_name: str) -> Tuple[tf.Tensor, tf.Tensor]:
+        """
+        Get the statistics for a given batch and layer.
+
+        Args:
+            batch_stats_holder (BatchStatsHolder): batch stats holder.
+            layer_name (str): the name of the layer.
+
+        Returns:
+            Tuple[tf.Tensor, tf.Tensor]: the mean and second moment for the specified batch and layer.
+        """
+        mean = batch_stats_holder.get_mean(bn_layer_name=layer_name)
```

Comment:
This should give the statistics on all batches not a specific batch. For a specific batch you have "batch_stats_holder"